File: Backend/services/api-gateway/src/grpc_impl/ctp.rs
// ============== CTP 服务 gRPC 调用封装 ==============

impl CtpGrpcClient {
    pub async fn receive_device_data(&mut self, device_no: String, data_type: i32, voltage: Option<String>, status_one: Option<String>, status_two: Option<String>, data_time: Option<String>) -> Result<grpc_proto::ctp::CtpResponse, tonic::Status> {
        Ok(self.inner().clone().receive_device_data(tonic::Request::new(grpc_proto::ctp::DeviceDataUpload { device_no, data_type, voltage, status_one, status_two, data_time })).await?.into_inner())
    }
    pub async fn control_lock(&mut self, factory_id: String, device_no: String, cmd_type: String, data: Option<String>) -> Result<grpc_proto::ctp::CommandResult, tonic::Status> {
        Ok(self.inner().clone().control_lock(tonic::Request::new(grpc_proto::ctp::LockControlRequest { factory_id, device_no, cmd_type, data })).await?.into_inner())
    }
    pub async fn get_device(&mut self, device_no: String) -> Result<grpc_proto::ctp::LockDevice, tonic::Status> {
        Ok(self.inner().clone().get_device(tonic::Request::new(grpc_proto::ctp::GetDeviceRequest { device_no })).await?.into_inner())
    }
    pub async fn list_devices(&mut self, park_code: Option<String>, status: Option<String>, page: i32, page_size: i32) -> Result<grpc_proto::ctp::DeviceListResponse, tonic::Status> {
        Ok(self.inner().clone().list_devices(tonic::Request::new(grpc_proto::ctp::DeviceQueryRequest { park_code, status, page, page_size })).await?.into_inner())
    }
}

File: Frontend/apps/admin/src/api/ctp.ts
/**
 * @file ctp.ts
 * @description CTP 平板锁设备管理 API
 */

import { httpClient } from '@/utils/alova';
import { handleApiError } from '@/utils/apiErrorHandler';

// ==================== 类型定义 ====================

/** CTP 设备数据类型枚举 */
export type CtpDataType = 0 | 1 | 2 | 5 | 6 | 7 | 8;
// 0=降板 1=升板 2=心跳 5=入位 6=出位 7=复位 8=逃费

/** 锁状态 */
export type LockStatus = 'Locked' | 'Unlocked' | 'Offline' | 'Fault';

/** 锁控制命令 */
export type CmdType = 'up' | 'down' | 'syn';

/** 状态字解析结果 */
export interface StatusParsed {
  lock_state: string;
  left_coil: string;
  right_coil: string;
  battery_raw: number;
  reset_flag: boolean;
  heartbeat: boolean;
  alarm: boolean;
}

/** CTP 设备 */
export interface CtpDevice {
  id: string;
  device_no: string;
  factory_id: string;
  status: LockStatus;
  battery: string | null;
  signal: number | null;
  voltage: string | null;
  park_code: string;
  created_at: string;
  updated_at: string;
  status_parsed?: StatusParsed;
  status_two_parsed?: {
    total_entry_count: number;
    total_exit_count: number;
    total_theft_count: number;
    current_occupancy: number;
  };
  battery_level?: string;
}

/** 设备上报数据 */
export interface DeviceDataUpload {
  device_no: string;
  data_type: CtpDataType;
  voltage?: string;
  status_one?: string;
  status_two?: string;
  data_time?: string;
}

/** 锁控制请求 */
export interface LockControlRequest {
  factory_id: string;
  device_no: string;
  cmd_type: CmdType;
  data?: string;
}

/** 设备查询参数 */
export interface DeviceQueryParams {
  park_code?: string;
  status?: LockStatus;
  page?: number;
  page_size?: number;
}

// ==================== API 函数 ====================

/** 接收设备上报数据 */
export async function reportDeviceData(data: DeviceDataUpload) {
  try {
    return await httpClient.post('/v1/ctp/report', data);
  } catch (error) {
    handleApiError(error, 'CTP设备管理');
    throw error;
  }
}

/** 发送锁控制命令 */
export async function controlLock(cmd: LockControlRequest) {
  try {
    return await httpClient.post('/v1/ctp/device/control', cmd);
  } catch (error) {
    handleApiError(error, 'CTP设备管理');
    throw error;
  }
}

/** 查询单个设备状态 */
export async function getDevice(deviceNo: string) {
  try {
    return await httpClient.get(`/v1/ctp/device/${deviceNo}`);
  } catch (error) {
    handleApiError(error, 'CTP设备管理');
    throw error;
  }
}

/** 分页获取设备列表 */
export async function listDevices(params?: DeviceQueryParams) {
  try {
    return await httpClient.post('/v1/ctp/device/list', { params });
  } catch (error) {
    handleApiError(error, 'CTP设备管理');
    throw error;
  }
}
